Sources/LeafKit/LeafData/LeafDataStorage.swift

```diff
@@ -3,46 +3,220 @@
+            case .int(_)             : return .int
+            case .string(_)          : return .string
+            // Internal Wrapped Types
+            case .lazy(_, let t, _),
+                 .optional(_, let t) : return t
+        }
+    }
+    
+    // MARK: Functions
+    
+    /// Will resolve anything but variant Lazy data (99% of everything), and unwrap optionals
+    internal func softResolve() -> LeafDataStorage {
+        guard invariant else { return self }
+        switch self {
+            case .lazy(let f, _, _): return f().storage
+            case .optional(let o, _):
+                if let unwrapped = o { return unwrapped }
+                return self
+            case .array(let a):
+                let resolved: [LeafData] = a.map {
+                    LeafData($0.storage.softResolve())
+                }
+                return .array(resolved)
+            case .dictionary(let d):
+                let resolved: [String: LeafData] = d.mapValues {
+                    LeafData($0.storage.softResolve())
+                }
+                return .dictionary(resolved)
+            default: return self
+        }
+    }
+
+    /// Will resolve anything but variant Lazy data (99% of everything)
+    internal func hardResolve() throws -> LeafData {
+        let resolveStored: LeafDataStorage = self.softResolve()
+        return LeafData(resolveStored)
+    }
+    
+    /// Will serialize anything to a String except Lazy -> Lazy
+    internal func serialize() throws -> String? {
+        let c = LeafConfiguration.self
+        switch self {
+            // Atomic non-containers
+            case .bool(let b)        : return c.boolFormatter(b)
+            case .int(let i)         : return c.intFormatter(i)
+            case .double(let d)      : return c.doubleFormatter(d)
+            case .string(let s)      : return c.stringFormatter(s)
+            // Data
+            case .data(let d)        : return c.dataFormatter(d)
+            // Wrapped
+            case .optional(let o, _) :
+                guard let wrapped = o else { return c.nilFormatter() }
+                return try wrapped.serialize()
+            // Atomic containers
+            case .array(let a)       :
+                let result = try a.map { try $0.storage.serialize() ?? c.nilFormatter() }
+                return c.arrayFormatter(result)
+            case .dictionary(let d)  :
+                let result = try d.mapValues { try $0.storage.serialize() ?? c.nilFormatter()}
+                return c.dictFormatter(result)
+            case .lazy(let f, _, _)  :
+                guard let result = f() as LeafData?,
+                      !result.storage.isLazy else { throw "Lazy can't return Lazy" }
+                return try result.storage.serialize() ?? c.nilFormatter()
+        }
+    }
+    
+    /// Final serialization to a shared buffer
+    internal func serialize(buffer: inout ByteBuffer) throws {
```

Comment:
Final serialization. The above serialize to String, run through configured encoding.